spec/integration/worker_async_spec.rb

```diff
@@ -28,51 +36,73 @@ module Backend
-        # For this test, we are going to perform a number of reports, and then,
-        # verify that the stats usage keys have been updated correctly.
+        worker = Worker.new(async: true, job_fetcher: JobFetcher.new(fetch_timeout: 1))
+
+        process_jobs(worker, n_reports, process_jobs_timeout_secs)
+
+        expect(Storage.instance.get(stats_key).to_i).to eq n_reports
+      end
+
+      context 'when there is a connection issue while trying to fetch jobs from the queue' do
+        let(:test_redis) { double }
 
-        n_reports = 10
+        before do
+          allow(test_redis).to receive(:blpop).and_raise(Errno::EPIPE)
+          allow(Backend.logger).to receive(:notify) # suppress errors in output
+        end
+
+        it 'recovers and continues processing jobs' do
+          # 'test_redis' will only be used once, after it raises, the worker
+          # will create a non-mocked instance of the client
+          worker = Worker.new(
+            async: true,
+            job_fetcher: JobFetcher.new(fetch_timeout: 1, redis_client: test_redis)
+          )
 
+          process_jobs(worker, n_reports, process_jobs_timeout_secs)
+
+          expect(Storage.instance.get(stats_key).to_i).to eq n_reports
+        end
+      end
+
+      def report_jobs(n_jobs, current_time, provider_key, service_id, app_id, metric_name, value)
         without_resque_spec do
           Timecop.freeze(current_time) do
-            n_reports.times do
+            n_jobs.times do
               Transactor.report(
-                  provider_key,
-                  service_id,
-                  0 => { app_id: app_id, usage: { metric_name => 1 } }
+                provider_key,
+                service_id,
+                0 => { app_id: app_id, usage: { metric_name => value } }
               )
             end
           end
         end
+      end
 
-        worker = Worker.new(async: true, job_fetcher: JobFetcher.new(fetch_timeout: 1))
+      def process_jobs(worker, n_jobs, timeout_seconds)
         worker_thread = Thread.new { worker.work }
 
-        stats_key = Stats::Keys.application_usage_value_key(
-            service_id, app_id, metric_id, Period[:day].new(current_time)
-        )
-
         # We do not know when the worker thread will finish processing all the
         # jobs. If it takes too much, we will assume that there has been some
         # kind of error.
         t_start = Time.now
 
-        while Storage.instance.get(stats_key).to_i < n_reports
-          if Time.now - t_start > 10
+        while Storage.instance.get(stats_key).to_i < n_jobs
+          if Time.now - t_start > timeout_seconds
             raise 'The worker is taking too much to process the jobs'
           end
 
           sleep(0.1)
         end
 
-        worker.shutdown
-
-        worker_thread.join
-
-        expect(Storage.instance.get(stats_key).to_i).to eq n_reports
+        ensure
+          worker.shutdown
+          worker_thread.join
```

Comment:
hm, I think you could just force MRI to kill the thread here if it did not finish in time to make sure we won't wait too much for whatever reason (ie. actually having some sort of connectivity problem, such as having the server respond very slowly), because as it stands we'll raise just fine, but we'll have to wait.

Reply:
I tried it without applying the fix and it worked fine.

Reply:
Writing a correct integration test for this is complicated because of some limitations of the testing suite that require some work to be fixed.
I'll merge this now because we need the fix to unblock other work, I'll try to come back to this next week.